Add Axelar bridge

npty opened the pull request:
# Description

This PR adds the `AxelarSenderAdapter` and `AxelarReceiverAdapter` contracts to the Multi-Message Aggregation (MMA) project, allowing for cross-chain message passing between different blockchain networks using the Axelar bridge. The deployment script for the Axelar adapters can be found at this [repository](https://github.com/axelarnetwork/axelar-multibridge).

Note: I understand that the `EIP5164` is still working in progress. So, I mark this PR as `draft` until the `EIP5164` branch gets merged into main branch.

0xkydo commented:
Thank you so much for the amazing work. We should merge `EIP5164` into `main` shortly.

0xkydo commented:
hey @npty, thank you for your patience. we are in the process of finalizing the implementation details of 5164. could you make this pr point to the EIP branch so we can do a bit more testing there?

Also, I would love to just double-check two things:
- you have implemented this portion from the EIP inside the receiver.
>> MessageExecutors MUST append the ABI-packed (messageId, fromChainId, from) to the calldata for each message being executed. This allows the receiver of the message to verify the cross-chain sender and the chain that the message is coming from.
- name variables as if this adapter will be used for other purposes outside MMA (aka, name srcSender instead of multimessageSender)


Thank you so much!

npty commented:
> hey @npty, thank you for your patience. we are in the process of finalizing the implementation details of 5164. could you make this pr point to the EIP branch so we can do a bit more testing there?
> 
> Also, I would love to just double-check two things:
> 
> * you have implemented this portion from the EIP inside the receiver.
> 
> > > MessageExecutors MUST append the ABI-packed (messageId, fromChainId, from) to the calldata for each message being executed. This allows the receiver of the message to verify the cross-chain sender and the chain that the message is coming from.
> 
> * name variables as if this adapter will be used for other purposes outside MMA (aka, name srcSender instead of multimessageSender)
> 
> Thank you so much!

Hi @0xkydo, thanks for getting back to me.

I've updated the pull request to point to the latest commit on the main branch, since I noticed that the EIP5164 branch has already been merged into it.

Regarding your two points:

1. Yes, I've implemented the portion from the EIP inside the receiver adapter and ensured that the calldata is ABI-packed with (messageId, fromChainId, from) for each message being executed.

2. I'll revise the variable names to make them more generic, so that the adapter can be used for other purposes outside of MMA. 


Thanks for your help and guidance throughout this process. Let me know if you have any further feedback or concerns.

0xkydo commented on contracts/adapters/axelar/AxelarSenderAdapter.sol:
```diff
@@ -0,0 +1,222 @@
+    uint32 public nonce;
+
+    // multiBridgeSender address
+    IMultiBridgeSender public multiBridgeSender;
```
This can be removed. The senderAdapter does not check where the message comes from but only relays it.

0xkydo commented on contracts/adapters/axelar/AxelarSenderAdapter.sol:
```diff
@@ -0,0 +1,222 @@
+    event AxelarChainRegistryUpdated(address axelarChainRegistry);
+    event GasLimitUpdated(uint32 gasLimit);
+
+    modifier onlyMultiBridgeSender() {
```
This is not needed.

ClaudeZsb commented on contracts/adapters/Celer/CelerSenderAdapter.sol:
```diff
@@ -4,6 +4,7 @@ pragma solidity 0.8.17;
 
 import "./interfaces/IMessageBus.sol";
 import "../../interfaces/IBridgeSenderAdapter.sol";
+import "../../interfaces/IMultiMessageReceiver.sol";
```
unused interface.

ClaudeZsb commented on contracts/adapters/axelar/AxelarSenderAdapter.sol:
```diff
@@ -0,0 +1,222 @@
+        require(msg.value >= fee, "insufficient fee");
+
+        // generate message id
+        bytes32 msgId = bytes32(uint256(nonce));
```
Could you please update to inherit from `../base/BaseSenderAdapter.sol` which provides a common way for generating `msgId`?

ClaudeZsb commented:
Basically, we need to update the `SenderAdapter` to be sender-agnostic.
Besides, I noticed that `AxelarGasService` has a gas refund mechanism. In current implementation, this refund goes to `MultiMessageSender.caller`. But after updating to be sender-agnostic, `AxelarSenderAdapter` would not know the right `fundAddress`. And remind that `MultiMessageSender` is not payable.
@npty  In a case where a contract sends message through Axelar without giving `fundAddress`, what's the best practice to deal with those gas left.

ermyas commented on contracts/adapters/axelar/AxelarReceiverAdapter.sol:
```diff
@@ -0,0 +1,116 @@
+        ) = abi.decode(payload, (bytes32, address, address, bytes));
+
+        // Check if the message has been executed
+        if (executeMsgs[msgId]) revert MessageIdAlreadyExecuted(msgId);
```
According to the [spec](https://github.com/axelarnetwork/cgp-spec/blob/306de046657e59f13dc822d37b2899d846e951a3/cgp-evm-v1.md?plain=1#L59) `validateContractCall()` should fail if a message is being re-executed. If so, is this separate check for whether a message is already executed ever really used?

ermyas commented on contracts/adapters/axelar/AxelarReceiverAdapter.sol:
```diff
@@ -0,0 +1,116 @@
+    IAxelarExecutable,
+    IBridgeReceiverAdapter
+{
+    mapping(uint256 => address) public senderAdapters;
```
For our purpose (i.e. for MMA V1 that is optimised for Uniswap's use case) we can maintain a single sender since the sender will always be Ethereum.

ermyas commented:
This is being integrated in PR #40
